### PR TITLE
Fix delete target table captured as input in event listener

### PR DESCRIPTION
**Pull request by chenjian2664**
## Description

Fix https://github.com/trinodb/trino/issues/16433

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# General
* Fix delete target table captured as input in event listener ({issue}`16433`)
```

### CODE REVIEW

**Review by Praveen2112 — COMMENTED** [review]
Can we add test for `DELETE` in `TestEventListenerBasic`.

**Comment by chenjian2664** [comment]
@Praveen2112 I am going to add the test cases in the `assertLineage`, is it reasonable, any suggestions?

**Comment by Praveen2112** [comment]
Yes and can we add a prep commit - which reproduces the issue (we could catch the failure) and then apply this commit to ensure that it works as expected.

**Comment by chenjian2664** [comment]
Sure, can, let me address it

**Comment by chenjian2664** [comment]
> Sure, can, let me address it

@Praveen2112 Added a test case to reproduce it, please review it

**Comment by chenjian2664** [comment]
https://github.com/trinodb/trino/actions/runs/4634226874/jobs/8202020666?pr=16887#step:6:34460 @Praveen2112

File: core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java [hunk 1]
```diff
@@ -83,6 +83,10 @@ public Set<Input> getInputs()
         @Override
         public Void visitTableScan(TableScanNode node, PlanFragmentId fragmentId)
         {
+//            if (node.isUpdateTarget()) {
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java:86]: left-over?

File: core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java [hunk 2]
```diff
@@ -83,6 +83,10 @@ public Set<Input> getInputs()
         @Override
         public Void visitTableScan(TableScanNode node, PlanFragmentId fragmentId)
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java:84]: please squash the commits and follow the guideline https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

File: core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java (hunk 1 above)
Comment (chenjian2664) [reply core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java]: Yes, it was intended to catch the error first.

File: core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java (hunk 2 above)
Comment (chenjian2664) [reply core/trino-main/src/main/java/io/trino/sql/planner/InputExtractor.java:84]: Please help to review again, thanks

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 3]
```diff
@@ -1222,6 +1222,21 @@ private void assertLineage(String baseQuery, Set<String> inputTables, OutputColu
         assertLineageInternal("CREATE VIEW mock.default.create_new_materialized_view AS " + baseQuery, inputTables, outputColumnMetadata);
         assertLineageInternal("INSERT INTO mock.default.table_for_output(test_varchar, test_bigint) " + baseQuery, inputTables, outputColumnMetadata);
         assertLineageInternal(format("DELETE FROM mock.default.table_for_output WHERE EXISTS (%s) ", baseQuery), inputTables);
+
+        // Test io input tables
+        assertIoInputTable("CREATE TABLE mock.default.create_io_input AS " + baseQuery, inputTables);
+        assertIoInputTable("INSERT INTO mock.default.insert_io_input(test_varchar, test_bigint) " + baseQuery, inputTables);
+        assertIoInputTable("DELETE FROM mock.default.table_for_io_input WHERE test_bigint IN (WITH s(test_varchar, test_bigint) AS (%s) SELECT test_bigint FROM s)".formatted(baseQuery), inputTables);
```
Comment (Praveen2112) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1229]: Can we use `table_for_output` to do the assertion here ?

Comment (chenjian2664) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1229]: Do you mean change the table name to `table_for_output`?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 4]
```diff
@@ -1222,6 +1222,21 @@ private void assertLineage(String baseQuery, Set<String> inputTables, OutputColu
...
+    }
+
+    private void assertIoInputTable(String sql, Set<String> inputTables)
+            throws Exception
+    {
+        QueryEvents queryEvents = runQueryAndWaitForEvents(sql).getQueryEvents();
+        QueryCompletedEvent event = queryEvents.getQueryCompletedEvent();
+        assertThat(event.getIoMetadata().getInputs())
+                .map(TestEventListenerBasic::getQualifiedName)
+                .containsExactlyInAnyOrderElementsOf(inputTables);
```
[the remark this answers is not in the document]
Comment (chenjian2664) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1239]: I am not sure we can make it. Note the  Line 1237 we are verify the `event.getIoMetadata().getInputs()` and the view creation won't have any input.
If can could you suggest how to use `assertLineageInternal` to verify?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 5]
```diff
@@ -1222,6 +1222,21 @@ private void assertLineage(String baseQuery, Set<String> inputTables, OutputColu
         assertLineageInternal("CREATE VIEW mock.default.create_new_materialized_view AS " + baseQuery, inputTables, outputColumnMetadata);
         assertLineageInternal("INSERT INTO mock.default.table_for_output(test_varchar, test_bigint) " + baseQuery, inputTables, outputColumnMetadata);
         assertLineageInternal(format("DELETE FROM mock.default.table_for_output WHERE EXISTS (%s) ", baseQuery), inputTables);
+
+        // Test io input tables
+        assertIoInputTable("CREATE TABLE mock.default.create_io_input AS " + baseQuery, inputTables);
+        assertIoInputTable("INSERT INTO mock.default.insert_io_input(test_varchar, test_bigint) " + baseQuery, inputTables);
```
Comment (Praveen2112) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1228]: These test are redundant - we have test for lineage assertion.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1239]: Can you elaborate more here ? We consider the query and run other operations on top of it.

[the remark this answers is not in the document]
Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1239]: Why don't we analyze `event.getMetadata().getTables()`

[the remark this answers is not in the document]
Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1239]: `QueryInputMetadata` returns almost all the input tables that are processed by a query and if the query has views - the view information is lost in `QueryInputMetadata` - so we treat `QueryMetadata#getTables` as a source of truth for lineage details.

[the remark this answers is not in the document]
Comment (chenjian2664) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:1239]: > Why don't we `analyze event.getMetadata().getTables()`

The `analyze event.getMetadata().getTables()` is different from the `event.getIoMetadata().getInputs()`, and the pr is for fixing the delete table occurs in the `event.getIoMetadata().getInputs()` see https://github.com/trinodb/trino/issues/16433

**Comment by chenjian2664** [comment]
https://github.com/trinodb/trino/actions/runs/5261427476/jobs/9509663329?pr=16887#step:6:25816 
Related to https://github.com/trinodb/trino/issues/16277 known issue

**Comment by mosabua** [comment]
:wave: @chenjian2664 @Praveen2112  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Pull request closed without merging**